### PR TITLE
Feature/recaptcha

### PR DESCRIPTION
**Pull request by jejelink69**

### CODE REVIEW

File: DependencyInjection/Configuration.php [hunk 1]
```diff
@@ -39,6 +39,7 @@ public function getConfigTreeBuilder()
                 ->append($this->addExtraFormTypesNode())
                 ->append($this->addExtraFormConstraintsNode())
                 ->append($this->addExtraFormConfiguratorsNode())
+                ->append($this->addReCaptchaConfiguration())
```
Comment (konandrum) [inline DependencyInjection/Configuration.php]: Node !

File: Form/Type/ReCaptchaType.php [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+     *
+     * @var RequestStack
+     */
+    private $requestStack;
+
+    /**
+     * @param array        $recaptchaParameters recaptchaParameters
+     * @param RequestStack $requestStack
+     */
+    public function __construct(array $recaptchaParameters, RequestStack $requestStack)
```
Comment (konandrum) [inline Form/Type/ReCaptchaType.php]: configuration/parameters

File: DependencyInjection/Configuration.php [hunk 3]
```diff
@@ -163,4 +164,42 @@ protected function addExtraFormConfiguratorsNode()
...
+                        ->scalarNode('port')->defaultValue(null)->end()
+                        ->scalarNode('auth')->defaultValue(null)->end()
+                    ->end()
+                ->end()
+                ->arrayNode('parameters')
+                    ->addDefaultsIfNotSet()
+                    ->children()
+                        ->booleanNode('enabled')->defaultTrue()->end()
+                        ->booleanNode('verify_host')->defaultFalse()->end()
+                        ->scalarNode('locale_key')->defaultValue('%kernel.default_locale%')->end()
```
Comment (konandrum) [inline DependencyInjection/Configuration.php]: Where it is used ?

File: Resources/config/constraints.yml [hunk 4]
```diff
@@ -328,3 +328,10 @@ idci_extra_form:
             extra_form_options:
                 value:   {"extra_form_type": "text", "options": {"required": true} }
                 message: {"extra_form_type": "text", "options": {"required": false} }
+
+        "recaptcha_validated":
```
Comment (konandrum) [inline Resources/config/constraints.yml]: verified

File: Validator/Constraints/ReCaptchaVerified.php [hunk 5]
```diff
@@ -0,0 +1,49 @@
+<?php
+
+namespace IDCI\Bundle\ExtraFormBundle\Validator\Constraints;
+
+use Symfony\Component\Validator\Constraint;
+
+/**
+ * @Annotation
+ * @Target("PROPERTY")
```
Comment (konandrum) [inline Validator/Constraints/ReCaptchaVerified.php:9]: ...

File: Validator/Constraints/ReCaptchaVerified.php [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+     */
+    public function __construct($privateKey)
+    {
+        $this->privateKey = $privateKey;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function getTargets()
```
Comment (konandrum) [inline Validator/Constraints/ReCaptchaVerified.php:29]: ?

File: Validator/Constraints/ReCaptchaVerifiedValidator.php [hunk 7]
```diff
@@ -0,0 +1,117 @@
...
+use Symfony\Component\HttpFoundation\RequestStack;
+use Symfony\Component\Security\Core\Authorization\AuthorizationChecker;
+use Symfony\Component\Security\Core\Authorization\AuthorizationCheckerInterface;
+use Symfony\Component\Validator\Constraint;
+use Symfony\Component\Validator\ConstraintValidator;
+
+class ReCaptchaVerifiedValidator extends ConstraintValidator
+{
+    /**
+     * Enable recaptcha?
```
Comment (konandrum) [inline Validator/Constraints/ReCaptchaVerifiedValidator.php]: remove ?

File: Validator/Constraints/ReCaptchaVerifiedValidator.php [hunk 8]
```diff
@@ -0,0 +1,117 @@
...
+     * @var array
+     */
+    protected $trustedRoles;
+
+    /**
+     * The reCAPTCHA verify server URL.
+     *
+     * @var string
+     */
+    protected $recaptchaVerifyServer;
```
Comment (konandrum) [inline Validator/Constraints/ReCaptchaVerifiedValidator.php]: url / server

File: Validator/Constraints/ReCaptchaVerifiedValidator.php [hunk 9]
```diff
@@ -0,0 +1,117 @@
...
+     * @param AuthorizationCheckerInterface|null $authorizationChecker
+     */
+    public function __construct(
+        array $recaptchaConfiguration,
+        RequestStack $requestStack,
+        AuthorizationCheckerInterface $authorizationChecker = null
+    ) {
+        $this->enabled = $recaptchaConfiguration['parameters']['enabled'];
+        $this->httpProxy = $recaptchaConfiguration['http_proxy'];
+        $this->recaptchaVerifyServer = 'https://'.$recaptchaConfiguration['parameters']['api_host'];
```
Comment (konandrum) [inline Validator/Constraints/ReCaptchaVerifiedValidator.php]: recaptchaVerifyUrl

File: Validator/Constraints/ReCaptchaVerified.php (hunk 5 above)
Comment (stacyharper) [reply Validator/Constraints/ReCaptchaVerified.php:9]: I think this force the constraint to be applied on properties and not on classes by example.

File: Form/Type/ReCaptchaType.php [hunk 10]
```diff
@@ -0,0 +1,156 @@
...
+
+    /**
+     * @param array        $configurationParameters configurationParameters
+     * @param RequestStack $requestStack
+     */
+    public function __construct(array $configurationParameters, RequestStack $requestStack)
+    {
+        $this->enabled = $configurationParameters['enabled'];
+        $this->apiHost = $configurationParameters['api_host'];
+        $this->recaptchaApiServer = sprintf('https://%s/recaptcha/api.js', $this->apiHost);
```
Comment (konandrum) [inline Form/Type/ReCaptchaType.php]: Url

File: ReCaptcha/RequestMethod/Post.php [hunk 11]
```diff
@@ -0,0 +1,75 @@
...
+
+    /**
+     * Constructor
+     *
+     * @param array $httpProxy proxy data to connect to
+     * @param string $recaptchaVerifyServer
+     */
+    public function __construct($recaptchaVerifyServer, array $httpProxy = [])
+    {
+        $this->recaptchaVerifyUrl = ($recaptchaVerifyServer ?: 'https://www.google.com').'/recaptcha/api/siteverify';
```
Comment (konandrum) [inline ReCaptcha/RequestMethod/Post.php]: recaptchaVerifyServer => recaptchaVerifyHost (without scheme)

File: Validator/Constraints/ReCaptchaVerified.php (hunk 5 above)
Comment (konandrum) [reply Validator/Constraints/ReCaptchaVerified.php:9]: This is the default behavior ?

Comment (stacyharper) [reply Validator/Constraints/ReCaptchaVerified.php:9]: The default behavior allow to apply the constraint on class and on property.

**Comment by stacyharper** [comment]
Fixed last comments. Ready to merge.

**Pull request merged**